crates/transpiler/src/passes/commutation_optimization.rs

```diff
@@ -0,0 +1,383 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::f64::consts::PI;
+
+use pyo3::prelude::*;
+use pyo3::{pyfunction, wrap_pyfunction, Bound, PyResult};
+use smallvec::smallvec;
+
+use crate::commutation_checker::CommutationChecker;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::operations::{radd_param, multiply_param, Operation, OperationRef, Param, StandardGate};
+use qiskit_circuit::packed_instruction::PackedInstruction;
+use qiskit_circuit::VarsMode;
+
+// the maximum number of qubits we check commutativity for
+const MAX_NUM_QUBITS: u32 = 3;
+
+const APPROXIMATION_DEGREE: f64 = 1.;
+
+/// RX and RZ rotation gates with smaller angles are considered as identity. 
```

Comment:
We have to be careful here: if we do any cutoff calculations we must ensure to use the same gate fidelity metric as the other passes. You should be able to re-use the functions used in `RemoveIdentityEquivalent`, I believe.

Reply:
Thanks! I have updated the pass to with the above in mind. Currently I have copied & adapted the code from `RemoveIdentityEquivalent`, the differences being:
* we do not support `target` for now; only `approximation_degree` (note that commutativity checker also does not support target),
* we need a py token (to handle merging of Pauli evolution gates),
* there is an option specifying the maximum number of qubits for matrix-based checks.
We should definitely think how to deduplicate the code. Ideally, we should port Pauli evolution gates to rust, and adjust the function that checks whether an instruction is equivalent to identity within the given tolerance inside `RemoveIdentityEquivalent` itself (thus allowing also this pass to remove small Pauli rotation gates).